lessons/babel.md

```diff
@@ -1 +1,130 @@
+  - JScript (IE < 9)
+- Some of the newer language features are unsupported in many of the currently used browsers
+- We want to be able to use modern features that make Javascript easier (and more fun) to develop in
+- Almost every browser that hits the Guardian supports the 'ES5' spec, Babel converts features from later specs back to ES5
+
+### What features does Babel handle?
+- Many of the improvements in the newer specs are syntatic sugar, which make transpilation relatively easy (default parameters, implicit returns)
+```js
+// ES2015
+const myFunc = (a = 1) => a + 2;
+
+// ES5 equivalent
+function myFunc(a) {
+  a = typeof a === "undefined" ? 1 : a;
+  return a + 2;
+}
+```
+- Some of the language features are new native APIs that are implementable (albeit less efficiently) in ES5 javascript and require a polyfill (e.g. Promise), this isn't handled by Babel transpilation! (Guardian uses polyfill.io [does it?!])
+- Some are not possible to completely polyfill (e.g. extending a built in class isn't possible in ES5) so they approximate functionality but may not work in all cases.
+
+## The project
+The app we're looking at revolves around a *tiny* Virtual DOM library (a pattern popular with libraries like React and Vue).
+
+The app make a request to CAPI and then uses the VDOM library to render the response the DOM and show a "Guardian" front page.
+
+We can test out the project now by running `yarn start` inside the project folder and navigating to the URL shown in the output. This will probably work fine in Chrome and Firefox but in Safari (at time of writing) it will have trouble with some of the syntax.
+
+## Getting Babel into a project
+This is the process to follow to get Babel into our project:
+- Install babel and it's CLI tool into the project
+- Configure babel: tell it what syntax we want to transform
+- Run babel on our code
+- Add in a polyfill for any missing APIs (babel doesn't handle these)
+
+### Installing Babel
+The Babel docs have pretty comprehensive info on this but it can sometimes be a lot to get your head around for the first time.
+
+- So first let's install babel into our project using our package manager
+  - We're using `yarn` so we need to run `yarn add --dev @babel/core @babel/cli` (for `npm i --save-dev @babel/core @babel/cli`)
+    - `@babel/cli`, as the name suggests, gives us a command line interfact to use babel
+    - `@babel/core` is the main babel code and `@babel/cli` will complain if we don't have it installed
+  - This will add both of those packages to our `node_modules` folder
+
+### Configurating Babel
+- We now have babel installed but we need to tell it the syntax we're using that we want to transpile
+- Babel transpiles *nothing* by default but instead uses plugins, we tell babel what to transpile by installing and specifying those plugins.
+- Babel, like a lot of javascript tooling, uses an "rc file" in the project root - by default this is `.babelrc` - for setting configuation
```

Comment:
`.babelrc.js`?

Reply:
@sndrs might be worth covering this. Do you have a use case for js in the rc file? I've only ever used the json syntax out of habit.